### PR TITLE
Hlt mvapuid

### PR DESCRIPTION
**Pull request by cvernier**
First version of HLT MVA Pu Id
@kirschen

### CODE REVIEW

**Comment by Martin-Grunewald** [comment]
What is so HLT specific about this that it needs to enter HLTrigger/JetMET, rather than the RecoJets subsystem?

**Comment by cvernier** [comment]
We are using HLT objects as inputs and we use the same subsystem as the other PU tool developed for HLT


@kkousour @kirschen

**Comment by ahinzmann** [comment]
What is the main difference between the offline PileupJetIdAlgo and the online MVAJetPuId? Is only the configuration different or also the c++ code?
Ideally one would like to share as much code as possible to easily track differences between online and offline.

**Comment by Martin-Grunewald** [comment]
Where? I believe you use standard Reco classes, so this is not HLT specific.
(HLT uses lots of reco objects, that does not make them HLT specific).
HLTrigger is typically for HLT filters, not (generic) producers,
and code which uses HLT specific C++ classes.
This looks more like it is generic reco which just happens
to be usefull for HLT as well.

**Comment by cvernier** [comment]
I understand your arguments but given that: a) the deadline is today and it will take more than 1 day to make the transition, b) the weights are stored in this subsystem, c) the associated tool made by Silvio (for pre-PF PU id) is in this subsystem, I propose to proceed as is.

@kkousour @kirschen

**Comment by kirschen** [comment]
Hi all,
ok, just a few words to set the scene:
This is the HLT-specific incarnation of the pileup jet id known from offline. However, it is not a plain retraining, just changing the collection names. Some of the input variables are different etc., so the code implementation is not same as for offline. From a JetMET-point of view we want to have the pileupId available in the online environment.
From my point of view, we should get it into the release now and, if it is deemed feasible and useful, merge the online/offline code base for the PileupJetId in the next release.

**Comment by Martin-Grunewald** [comment]
Sorry, 75X does not have any deadlines - or do you mean ORP deadlines? Those are weekly.
Also, understand that 75X is not yet relevant for data taking, as that happens with
74X, so a backport PR to 74X is needed anyway. And take the opportunity to move
Silvio's pre-PF PUId to the proper place (although that is much less heavy
than the MVA part you propose).

**Comment by slava77** [comment]
On 5/18/15 9:41 AM, Martin Grunewald wrote:
> Sorry, 75X does not have any deadlines - or do you mean ORP deadlines?

Today is the deadline for new features in 75X.
After that only bugfixes are supposed to be made.

File: HLTrigger/JetMET/plugins/MVAJetPuIdProducer.h [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+#include "DataFormats/JetReco/interface/PileupJetIdentifier.h"
+#include "DataFormats/VertexReco/interface/Vertex.h"
+
+#include "CondFormats/JetMETObjects/interface/JetCorrectorParameters.h"
+#include "CondFormats/JetMETObjects/interface/FactorizedJetCorrector.h"
+#include "JetMETCorrections/Objects/interface/JetCorrectionsRecord.h"
+
+#include "FWCore/ParameterSet/interface/FileInPath.h"
+
+class MVAJetPuIdProducer : public edm::EDProducer {
```
Comment (Martin-Grunewald) [inline HLTrigger/JetMET/plugins/MVAJetPuIdProducer.h]: Must be a stream or global EDProcucer for multi threading

File: HLTrigger/JetMET/plugins/MVAJetPuIdProducer.cc [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+         iEvent.put(idsout);
+     }
+ }
+ 
+ 
+ 
+ void
+ MVAJetPuIdProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+         edm::ParameterSetDescription desc;
+     desc.setUnknown();
```
Comment (Martin-Grunewald) [inline HLTrigger/JetMET/plugins/MVAJetPuIdProducer.cc]: This is not acceptable. The fillDescription method has to properly declare explicitly all parameters
which are read by the module plugin in its constructor and wherever/whatever else.

**Comment by Martin-Grunewald** [comment]
@slava77 
Ah, so leaving it to the last day - and then using the deadline as an argument....
Hmmmm.....

File: HLTrigger/JetMET/plugins/MVAJetPuIdProducer.cc [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+     jecCor_ = 0;
+ 
+     if( ! runMvas_ ) assert( algos.size() == 1 );
+     
+     if( produceJetIds_ ) {
+         produces<edm::ValueMap<StoredPileupJetIdentifier> > ("");
+     }
+     for(std::vector<edm::ParameterSet>::iterator it=algos.begin(); it!=algos.end(); ++it) {
+         std::string label = it->getParameter<std::string>("label");
+         algos_.push_back( std::make_pair(label,new MVAJetPuId(*it)) );
```
Comment (Martin-Grunewald) [inline HLTrigger/JetMET/plugins/MVAJetPuIdProducer.cc]: Where are these "new"s deleted? Otherwise there is a memory leak!

**Comment by slava77** [comment]
IIRC, the policy for feature changes in the closing/closed release is to
have the change integrated in the development release (there will be 76X
after 740pre5 comes out)
and then backport from the development release once it's validated.
This is possible for features that do not (noticeably) change physics
performance in terms of DQM/Validation.

It seems to me that the feature discussed here should just go to 76X
and then planned to be backported.

On 5/18/15 9:54 AM, Martin Grunewald wrote:
> @slava77 <https://github.com/slava77>
> Ah, so leaving it to the last day - and then using the deadline as an
> argument....
> Hmmmm.....
> 
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/9133#issuecomment-103085763>.
> 

-- 
***********************************************************************
 Vyacheslav (Slava) Krutelyov
 TAMU: Physics Dept Texas A&M MS4242, College Station, TX 77843-4242
 CERN: 42-R-027
 AIM/Skype: siava16              googleTalk: slava77@gmail.com
 (630) 291-5128  Cell (US)       +41 76 275 7116 Cell (CERN)

-----------------------------------------------------------------------

**Comment by fwyzard** [comment]
On 18 May 2015 at 17:43, Slava Krutelyov <notifications@github.com> wrote:

It seems to me that the feature discussed here should just go to 76X
> and then planned to be backported.


Note that this means *backported to 74x* for HLT data taking in 2015.

Ciao,
.Andrea

-- 
Strategy is a system of expedients.

Field Marshall Helmuth Carl Bernard Graf von Moltke

**Comment by cvernier** [comment]
Ok. Thanks. I will update the code as Martin suggested and make another PR for 76X. 
Please let me know which subsystem should be preferred: HLTrigger/JetMET or RecoJets 
cheers
Caterina

**Comment by slava77** [comment]
On 5/18/15 11:17 AM, Andrea Bocci wrote:
> On 18 May 2015 at 17:43, Slava Krutelyov <notifications@github.com> wrote:
> 
> It seems to me that the feature discussed here should just go to 76X
>> and then planned to be backported.
> 
> 
> Note that this means *backported to 74x* for HLT data taking in 2015.

I guess this puts a more specific time target for delivery.

Is this MVA PU ID going to be used in some test triggers as an investigation
to apply it for the majority of triggers, say, in 25 ns run?

**Comment by Martin-Grunewald** [comment]
I still believe the better place is RecoJets even and espcially if it is (or remains)
a variant of an offline code: this will help maintenance and aid in checking
whether they could/should be unified and whether changes to one should
apply to the other.

**Pull request closed without merging**